### PR TITLE
@/CivicSignalBlog - Add ability to manage CivicSignal content

### PR DESCRIPTION
**Pull request by m453h**
## Description
This PR enhances the CivicSignalBlog by adding the ability to manage CivicSignal [web-tools ](https://civicsignal.africa/#/home)content.

This implementation  allows users to switch between different instances of the web-tools which include: 
- Research Blog
- CivicSignal Site
- Explorer
- Topic Mapper
- Source Manager

This initial implementation only supports:
- Pages Collection: Allows managing basic pages specific to each tool.
- Global Settings: Includes global configuration options for each of the above applications.

This PR focuses on simplicity and paves the way for future enhancements.

## Type of change
- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## Screenshots

Screenshot 2024-09-20 at 17 08 02

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation

### CODE REVIEW

**Review by koechkevin — APPROVED** [review]
LGTM

File: apps/civicsignalblog/src/lib/data/json/applications.js [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+  "Explorer",
+  "Topic Mapper",
+  "Source Manager",
+  "Research Blog",
+].map((label) => ({
+  label,
+  value: label.toLowerCase().replace(/\s+/g, "_"),
+}));
+
+export default applications;
```
Comment (kilemensi) [inline apps/civicsignalblog/src/lib/data/json/applications.js:12]: Not sure why this is in `lib/data/json`... it's certainly not a JSON.

File: apps/civicsignalblog/src/lib/data/json/applications.js [hunk 2]
```diff
@@ -0,0 +1,12 @@
+const applications = [
+  "CivicSignal",
+  "Explorer",
+  "Topic Mapper",
+  "Source Manager",
+  "Research Blog",
+].map((label) => ({
+  label,
+  value: label.toLowerCase().replace(/\s+/g, "_"),
+}));
```
Comment (kilemensi) [inline apps/civicsignalblog/src/lib/data/json/applications.js:10]:
```js
export const MAIN = "main";
export const EXPLORER = "explorer";
export const TOPIC_MAPPER = "topic-mapper";
export const SOURCE_MAPPER = "source-manager";
export const RESEARCH = "research",

export default applications = {
  MAIN: "CivicSignal",
  EXPLORER: "Explorer",
  TOPIC_MAPPER: "Topic Mapper",
  SOURCE_MAPPER: "Source Manager",
  RESEARCH: "Research",
}
```

1. I think the above is much cleaner and allows `import { MAIN } from "applications"` enabling the slug to be reused without being retyped over and over i.e. single place to change if we ever need to.
2. I don't see **Media Data**

File: apps/civicsignalblog/tsconfig.json [hunk 3]
```diff
@@ -20,7 +20,13 @@
       "@/commons-ui/next/*": ["../../packages/commons-ui-next/src/*"]
     }
   },
-  "include": ["next-env.d.ts", "**/*.ts", "**/*.tsx"],
+  "include": [
+    "next-env.d.ts",
+    "**/*.ts",
+    "**/*.tsx",
+    "src/payload/components/BeforeDashboard/index.jsx",
+    "src/payload/components/BeforeDashboard/index.jsx"
```
Comment (kilemensi) [inline apps/civicsignalblog/tsconfig.json:28]: What does including/not including them here actually do?

File: apps/civicsignalblog/src/payload/collections/Users.js [hunk 4]
```diff
@@ -1,3 +1,6 @@
+import payload from "payload";
+
+import applications from "../../lib/data/json/applications";
```
Comment (kilemensi) [inline apps/civicsignalblog/src/payload/collections/Users.js:3]: Not sure I fully understand the naming of "managed applications"? Who is managing them and why are they managing them?

I think authorised applications/apps or just applications/apps would make a lot more sense (unless there is a reason behind the "managed" part)

File: apps/civicsignalblog/src/payload/collections/Users.js [hunk 5]
```diff
@@ -53,6 +56,104 @@ const Users = {
...
     },
+    {
+      name: "defaultManagedApplication",
+      type: "select",
+      hasMany: false,
+      admin: {
+        isClearable: true,
+        isSortable: true,
+      },
+      options: applications,
```
Comment (kilemensi) [inline apps/civicsignalblog/src/payload/collections/Users.js:67]: Should default to `research` app as that's what is already there.

**Review by kilemensi — COMMENTED** [review]
Looks really good... kudos

---

1. Isn't there a way of using URL for switch apps? Or at least show the current app as part of the URL i.e. `http://localhost:3000/admin/research` or `http://localhost:3000/admin?app=research` or _something_.
2. Lets move the app switch to where user/locale is.
Screenshot 2024-09-23 at 14 35 27

**Review by VinneyJ — APPROVED** [review]

File: apps/civicsignalblog/src/payload/collections/Users.js (hunk 5 above)
Comment (m453h) [reply apps/civicsignalblog/src/payload/collections/Users.js:68]: Makes sense I have made the change

File: apps/civicsignalblog/src/payload/collections/Users.js (hunk 4 above)
Comment (m453h) [reply apps/civicsignalblog/src/payload/collections/Users.js]: I kept "managed" since they are apps whose content is being managed by our CMS 😄 ... but come to think of it, its somewhat redundant...I think we can keep it as just "app"

File: apps/civicsignalblog/tsconfig.json (hunk 3 above)
Comment (m453h) [reply apps/civicsignalblog/tsconfig.json]: I don't even know how that got there 🤦 , must've been automatically added, I've removed it !

File: apps/civicsignalblog/src/lib/data/json/applications.js (hunk 2 above)
Comment (m453h) [reply apps/civicsignalblog/src/lib/data/json/applications.js]: I have refactored the app name to something very close to your suggestion, since we use the app names in the labels for the select field in the user collection I had to make a minor adjustment to get  the `{value: "", key:""}` array.

The **Media Data** is actually just a page under the MAIN (CivicSignal) its not a separate app that's why I didn't include it in the list

**Comment by m453h** [comment]
> Looks really good... kudos
> 
> 1. Isn't there a way of using URL for switch apps? Or at least show the current app as part of the URL i.e. `http://localhost:3000/admin/research` or `http://localhost:3000/admin?app=research` or _something_.
> 2. Lets move the app switch to where user/locale is.
> 
> Screenshot 2024-09-23 at 14 35 27

I've made the switch, we can show the URL based on the App the user has switched to when we redirect to the homepage but  we will loose it whenever we click any other link, the same behaviour is observed for the user locale. 
ezgif-1-c7d30d5837

**Comment by kilemensi** [comment]
> I've made the switch, we can show the URL based on the App the user has switched to when we redirect to the homepage but we will loose it whenever we click any other link ...
>

Cool @m453h . It's minor so we can leave it out for now since everything is CivicSignal.

File: apps/civicsignalblog/src/lib/data/json/applications.js (hunk 1 above)
Comment (kilemensi) [reply apps/civicsignalblog/src/lib/data/json/applications.js:22]: Has this been addressed @m453h ?

File: apps/civicsignalblog/src/payload/access/canAccessApplication.js [hunk 6]
```diff
@@ -0,0 +1,10 @@
+export default function canAccessApplication(user, searchString) {
+  if (user) {
+    const app = user.currentApp || user.defaultApp;
+    if (app) {
+      return app === searchString.toLowerCase();
+    }
+    return false;
+  }
+  return false;
```
Comment (kilemensi) [inline apps/civicsignalblog/src/payload/access/canAccessApplication.js:9]:
```suggestion
  if (user) {
    const app = user.currentApp || user.defaultApp;
    return app === searchString.toLowerCase();
  }
  return false;
```

Don't think we need `if`/multiple `return`s

File: apps/civicsignalblog/src/payload/collections/CivicSignalBlog/Authors.js [hunk 7]
```diff
@@ -1,7 +1,9 @@
+import canRead from "../../access/applications/researchBlog";
```
Comment (kilemensi) [inline apps/civicsignalblog/src/payload/collections/CivicSignalBlog/Authors.js:1]:
```suggestion
import canRead from "../../access/applications/research";
```

Similarly for the other apps, make them match what we've defined in `applications.js`

File: apps/civicsignalblog/src/payload/collections/CivicSignalBlog/Pages.js [hunk 8]
```diff
@@ -0,0 +1,31 @@
+import canRead from "../../access/applications/researchBlog";
```
Comment (kilemensi) [inline apps/civicsignalblog/src/payload/collections/CivicSignalBlog/Pages.js:1]: Can't we get aliases to work in Payload? Not necessary in this PR but these relative paths are hard to keep up with.

File: apps/civicsignalblog/src/payload/collections/CivicSignalBlog/Pages.js [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+import CustomPageHeader from "../../blocks/CustomPageHeader";
+import Error from "../../blocks/Error";
+import FeaturedStories from "../../blocks/FeaturedStories";
+import LongForm from "../../blocks/LongForm";
+import PageHeader from "../../blocks/PageHeader";
+import Posts from "../../blocks/Posts";
+import pages from "../../fields/pages";
```
Comment (kilemensi) [inline apps/civicsignalblog/src/payload/collections/CivicSignalBlog/Pages.js:8]: I don't think `pages` is a field... may be just a function in `utils`?

File: apps/civicsignalblog/src/payload/collections/Users.js [hunk 10]
```diff
@@ -53,6 +56,76 @@ const Users = {
...
+    },
+  ],
+  endpoints: [
+    {
+      path: "/current-app",
```
Comment (kilemensi) [inline apps/civicsignalblog/src/payload/collections/Users.js:84]: Minor but
```suggestion
      path: "/apps/current",
```
may be a bit more RESTy? (or `users/me/apps/current`)

File: apps/civicsignalblog/src/payload/collections/Users.js [hunk 11]
```diff
@@ -53,6 +56,76 @@ const Users = {
...
+
+        const updatedUser = await payload.update({
+          collection: "users",
+          id: userId,
+          data: {
+            currentApp: selectedApp || currentUser.defaultApp,
+          },
+        });
+
+        res.status(200).send({
```
Comment (kilemensi) [inline apps/civicsignalblog/src/payload/collections/Users.js:123]:
```suggestion
        }

        const { selectedApp } = req.body;
        if (!selectedApp) {
          res.status(400).send({
            error: `Incorrect message format was received:${JSON.stringify(req.body)}`,
          });
        }

        const userId = req.user.id;
        const currentUser = await payload.findByID({
          collection: "users",
          id: userId,
          showHiddenFields: true,
        });
        if (!currentUser) {
          res
            .status(404)
            .send({ error: "User with specified ID was not found" });
        }

        const updatedUser = await payload.update({
          collection: "users",
          id: userId,
          data: {
            currentApp: selectedApp || currentUser.defaultApp,
          },
        });
        res.status(200).send({
```

Too many empty lines make it harder to follow which parts of code are related/dependant.

File: apps/civicsignalblog/src/payload/components/actions/index.tsx [hunk 12]
```diff
@@ -0,0 +1,98 @@
+import { useAuth } from "payload/components/utilities";
+import React, { useEffect, useState } from "react";
+
+import applications from "../../../lib/data/json/applications";
+
+function BeforeDashboard() {
+  const { user } = useAuth();
+
+  const [selectedApp, setSelectedApp] = useState(
+    String(user.currentApp ?? user.defaultApp) ?? "",
```
Comment (kilemensi) [inline apps/civicsignalblog/src/payload/components/actions/index.tsx:10]:
```suggestion
    user.currentApp || user.defaultApp || RESEARCH,
```

I'd rather we use really defaults whenever we can instead of `""`. Alternatively, ensure user.defaultApp is never undefined/null/empty.

File: apps/civicsignalblog/src/payload/globals/Site/researchBlog.js [hunk 13]
```diff
@@ -0,0 +1,18 @@
+import canRead from "../../access/applications/researchBlog";
+import settings from "../../fields/settings";
+
+import EngagementTab from "./Components/EngagementTab";
+import GeneralTab from "./Components/GeneralTab";
+import NavigationTab from "./Components/NavigationTab";
+
+const ResearchBlog = settings({
```
Comment (kilemensi) [inline apps/civicsignalblog/src/payload/globals/Site/researchBlog.js:8]:
```suggestion
const Research = settings({
```

File: apps/civicsignalblog/src/lib/data/json/applications.js (hunk 1 above)
Comment (m453h) [reply apps/civicsignalblog/src/lib/data/json/applications.js:22]: @kilemensi, I think much like `lib/data/json/countries.js` (which is currently not being referenced anywhere), I propose to move them to the `lib/data/common`

File: apps/civicsignalblog/src/payload/collections/CivicSignalBlog/Pages.js (hunk 8 above)
Comment (m453h) [reply apps/civicsignalblog/src/payload/collections/CivicSignalBlog/Pages.js:1]: Yeah that was the plan (I didn't do it in this PR to reduce the number of files changed)... working with relative paths has been a pain

File: apps/civicsignalblog/src/payload/collections/CivicSignalBlog/Pages.js (hunk 9 above)
Comment (m453h) [reply apps/civicsignalblog/src/payload/collections/CivicSignalBlog/Pages.js]: Done I have moved it to utils

File: apps/civicsignalblog/src/lib/data/json/applications.js (hunk 1 above)
Comment (m453h) [reply apps/civicsignalblog/src/lib/data/json/applications.js]: Decided to move it to payload/lib/data/common

File: apps/civicsignalblog/src/payload/collections/Research/Pages.js [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+import CustomPageHeader from "../../blocks/CustomPageHeader";
+import Error from "../../blocks/Error";
+import FeaturedStories from "../../blocks/FeaturedStories";
+import LongForm from "../../blocks/LongForm";
+import PageHeader from "../../blocks/PageHeader";
+import Posts from "../../blocks/Posts";
+import pages from "../../utils/createPagesCollection";
+
+const Pages = pages({
+  pageSlug: "pages",
```
Comment (kilemensi) [inline apps/civicsignalblog/src/payload/collections/Research/Pages.js:11]: Minor and hence can be ignored or implemented in another PR but I think it makes a bit more sense to migrate `pageSlug` from `pages` to `${RESEARCH}-pages` now when we have have very few pages in PROD

File: apps/civicsignalblog/src/payload/components/actions/index.tsx [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+            throw new Error("Network response was not ok");
+          }
+        })
+        .then((data) => {
+          window.location.replace(`/admin?app=${encodeURI(data.currentApp)}`);
+        })
+        .catch((error) => {
+          setLoading(false);
+          throw error;
+        });
```
Comment (kilemensi) [inline apps/civicsignalblog/src/payload/components/actions/index.tsx:43]: Minor but shouldn't we be using `async`? If we **_have to_** use promise, shouldn't there just be a single `finally` that sets loading to false?

File: apps/civicsignalblog/src/payload/components/actions/index.tsx [hunk 16]
```diff
@@ -0,0 +1,98 @@
...
+    }
+  }, [selectedApp]);
+
+  const handleChange = (event) => {
+    const selectedValue = event.target.value;
+    setSelectedApp(selectedValue);
+  };
+
+  return (
+    <div style={{}}>
```
Comment (kilemensi) [inline apps/civicsignalblog/src/payload/components/actions/index.tsx:53]:
```suggestion
    <div>
```

File: apps/civicsignalblog/src/payload/components/actions/index.tsx [hunk 17]
```diff
@@ -0,0 +1,98 @@
...
+          <p
+            style={{
+              padding: "5px 10px",
+              marginTop: "30px",
+              background: "var(--theme-input-bg)",
+            }}
+          >
+            Loading...
+          </p>
+        </>
```
Comment (kilemensi) [inline apps/civicsignalblog/src/payload/components/actions/index.tsx:65]:
```suggestion
        <p
          style={{
            padding: "5px 10px",
            marginTop: "30px",
            background: "var(--theme-input-bg)",
          }}
        >
          Loading...
        </p>
```

File: apps/civicsignalblog/src/payload/globals/Site/research.js [hunk 18]
```diff
@@ -0,0 +1,18 @@
+import canRead from "../../access/applications/research";
+import settings from "../../utils/createGlobalSettings";
+
+import EngagementTab from "./Components/EngagementTab";
+import GeneralTab from "./Components/GeneralTab";
+import NavigationTab from "./Components/NavigationTab";
+
+const Research = settings({
+  slug: "settings-site",
```
Comment (kilemensi) [inline apps/civicsignalblog/src/payload/globals/Site/research.js:9]: Same here, might be best to migrate to `settings-${RESEARCH}-site`

**Review by kilemensi — APPROVED** [review]
🚀

File: apps/civicsignalblog/src/payload/collections/Research/Pages.js (hunk 14 above)
Comment (m453h) [reply apps/civicsignalblog/src/payload/collections/Research/Pages.js]: ✅ 

I've done it now, I think we'll  have to recreate the pages when we migrate to prod, but its a very minor issue

Comment (kilemensi) [reply apps/civicsignalblog/src/payload/collections/Research/Pages.js]: ...or a [migration](https://payloadcms.com/docs/database/migrations) script instead of recreating them.
